### PR TITLE
ruby27: Add begin-less ranges support

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

New try on the begin-less ranges (https://github.com/sorbet/sorbet/pull/3260).

The problem raised by @elliottt with:

```ruby
low = T.unsafe(0)
high = T.unsafe(16)
T.reveal_type(low..high) # T::Range[T.noreturn], but T::Range[T.untyped] on master
```

Was happening because with the proposed signatures for `Range.new` the parameter type could not resolve when one of the end was `NilClass` and Sorbet used `bottom`.

This pull-request adds signatures resolving with `NilClass` so we can get the `T::Range[T.untyped]` as expected.

Regarding the case `Range(nil, nil)`, the [original idea](https://github.com/sorbet/sorbet/pull/3260#discussion_r451028410) was to use `T.noreturn` as to raise errors on calls to `first`, `last` and `each` because since Ruby 2.7:

```
irb(main):001:0> r = Range.new(nil, nil)
=> nil..nil
irb(main):002:0> r.first
=> RangeError (cannot get the first element of beginless range)
irb(main):003:0> r.last
=> RangeError (cannot get the last element of endless range)
irb(main):004:0> r.each
=> #<Enumerator: nil..nil:each>
```

But it turns out this might no be compatible with previous versions of Ruby:

```
irb(main):001:0> r = Range.new(nil, nil)
=> nil..
irb(main):002:0> r.first
=> nil
irb(main):003:0> r.last
=> RangeError (cannot get the last element of endless range)
irb(main):004:0> r.each
=> #<Enumerator: nil..:each>
```

Only `last` raises an error with 2.6 and `each` return an enumerator over `nil` in both version so I don't think it makes sense to type it as `T::Range[T.noreturn]`. The changes in this pull-request makes it a `T::Range[T.noreturn]`.

### Test plan

I added a bunch of examples in the `range.rb` test.

### CODE REVIEW

File: test/testdata/desugar/range.rb [hunk 1]
```diff
@@ -14,9 +14,39 @@ def foo
...
+    g = (..1)
+    h = Range.new(nil, 1)
     T.reveal_type(g) # error: Revealed type: `T::Range[Integer(1)]`
     T.reveal_type(h) # error: Revealed type: `T::Range[Integer(1)]`
+
+    # testing for endless ranges
+    i = (1..)
+    j = Range.new(1, nil)
+    T.reveal_type(i) # error: Revealed type: `T::Range[Integer(1)]`
+    T.reveal_type(j) # error: Revealed type: `T::Range[Integer(1)]`
```
Comment (elliottt) [inline test/testdata/desugar/range.rb:27]: This doesn't seem quite right to me, The range will produce values other than `1`, right?

I'm not entirely sure if this is something that we can fix with an rbi change, so if you can't fix it it might be easier to look at adding an intrinsic in `core/types/calls.cc`.

File: test/testdata/desugar/range.rb [hunk 2]
```diff
@@ -14,9 +14,39 @@ def foo
...
+
+    # testing nil..nil edge case
+    k = (nil..)
+    l = Range.new(nil, nil)
+    T.reveal_type(k) # error: Revealed type: `T::Range[NilClass]`
+    T.reveal_type(l) # error: Revealed type: `T::Range[NilClass]`
```
Comment (paracycle) [inline test/testdata/desugar/range.rb:33]: I think both of these should be `T::Range[T.untyped]` since we can't say anything about the entries in the range. The fact that the boundaries are `nil` does not mean that all the entries are `nil`, it just means there are no boundaries.

File: test/testdata/desugar/range.rb (hunk 1 above)
Comment (Morriar) [reply test/testdata/desugar/range.rb:27]: Oh you're right, I didn't realized it was using `1` everywhere.

Yeah I think at this point I reached the limit of the RBI power. I'll add a special case for ranges.

File: test/testdata/desugar/range.rb (hunk 2 above)
Comment (Morriar) [reply test/testdata/desugar/range.rb:33]: Can it?

Do you refer to things like `T::Range[Integer].new(nil, nil)`? On both 2.6 and 2.7 I was unable to get any other value out of this kind of range but `nil`.

**Comment by Morriar** [comment]
@elliottt I added an intrinsic for `Range::new` and corrected the test output 👍

**Comment by elliottt** [comment]
> @elliottt I added an intrinsic for `Range::new` and corrected the test output 👍

I noticed that there's a signature for `Range.new` and `Range#initialize` in the rbi, could you remove the `Range.new` sig if it's not needed?

Also, do your changes work if the intrinsic is for `Range#initialize` instead? I'd like to avoid not using the existing intrinsic for `new` if possible; it will use `initialize` for getting argument type information.

File: test/testdata/desugar/range.rb (hunk 2 above)
Comment (paracycle) [reply test/testdata/desugar/range.rb:34]: Here is why I think this might lead to type problems. If I have a database query interface that works with ranges, I can define it like this:
```ruby
class Model
  sig { params(id_range: T::Range[Integer]).returns(T::Array[Model]) }
  def self.find(id_range)
    # do some querying
  end
end
```

with the intention of it being used in the following ways:
```ruby
# Query a range of ids
Model.find(1..3) #=> [Model(1), Model(2), Model(3)]

# Query ids after 5
Model.find(5..) #=> [Model(5), Model(6)]

# Query ids up to 2
Model.find(..2) #=> [Model(1), Model(2)]

# Query all records
Model.find(Range.new(nil, nil)) #=> [Model(1), Model(2), Model(3), Model(4), Model(5), Model(6)]
```

However, if Sorbet infers the type of `Range.new(nil, nil)` as `T::Range[NilClass]` the last call above will not type-check since `T::Range[NilClass]` will not be compatible with `T::Range[Integer]`. But that does not sound right!

That is why I am claiming that (regardless of what items we can get out of it), we should treat unbounded ranges as items that we have no idea what is inside, so that it is compatible with all interfaces. 

Thus, I propose this should be `T::Range[T.untyped]`

Comment (elliottt) [reply test/testdata/desugar/range.rb:34]: I agree, and I think that with the changes that @Morriar and I discussed last week, it will be.

**Comment by Morriar** [comment]
@elliottt this has been rebased on master since #3359 was merged.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HzdazFYuxZsTZv

**Review by jez — APPROVED** [review]
Passes on Stripe's code base!

File: test/testdata/desugar/range.rb [hunk 3]
```diff
@@ -62,6 +62,18 @@ def foo
...
+
+  lr21 = (..nil)
+  T.reveal_type(lr21) # error: Revealed type: `T::Range[T.untyped]`
+
+  lr22 = (..untyped)
+  T.reveal_type(lr22) # error: Revealed type: `T::Range[T.untyped]`
+
+  lr23 = (..nilable)
+  T.reveal_type(lr23) # error: Revealed type: `T::Range[Integer]`
+
```
Comment (jez) [inline test/testdata/desugar/range.rb:76]: 👍 thanks!
